### ISSUE TITLE
Publish SHA256 hashes for binaries on the download page

### ISSUE THREAD

**Issue by voltagex** [mholt/caddy#1658]
Related to https://twitter.com/mholt6/status/861824879481548800

Especially on Windows, PGP can be tricky to set up. A SHA256 hash should be verifyable on all operating systems without extra software.

**Comment by slightfoot** [comment]
I use HashTab for this sort of thing, it might be worth mentioning a couple on the page perhaps with a non-endorsement disclaimer.

> http://implbits.com/products/hashtab/

**Comment by JoshHarmon** [comment]
How do these hashes provide any assurance of authenticity? All the hash matching tells you is that you downloaded a file that wasn't corrupted in transit. However, TCP and TLS prevent corruption (via lost packets) and MITM tampering when downloading the releases. To me, this seems rather pointless.

The only way you get authenticity from a hash is with using an HMAC construction, which involves adding a _secret_ key into the mix (HMACs are for symmetric encryption, which we aren't after here), which would make no sense here (as everyone would need the key to verify the hash, but then everyone knows the _secret_ key). Is there something I'm missing?

**Comment by mholt** [comment]
I view this as a shortcoming in Windows more than the Caddy website. As Josh explained, simply showing a hash (in our case) isn't really going to provide more security. In order to compromise a build in the cache, you'd need write access to the disk, but if you have write access to the disk, you might as well change the hash displayed to users while you're at it. (Something like this [already happened to Transmission](https://transmissionbt.com/keydnap_qa/).)

If we showed merely unsigned hashes, we would have to do so in a distributed way: on GitHub, on the Caddy website, at the very least. But to show hashes for custom builds (builds with plugins, i.e. not the assets on GitHub) anywhere but the Caddy website would be nigh impossible. Even if we did distribute the hashes to other sites, we have to ensure that any and all changes to the hashes are authorized. [Homebrew recently distributed a malicious version of Handbrake](https://twitter.com/mholt6/status/861067314048802816) by simply accepting a hash change without verification.

It's a hard problem.

I'm sorry it's difficult to verify signatures on Windows (and I hate PGP, even on unix), but I don't think simply showing hashes on the Caddy site is going to provide any of the assurances you would expect. Perhaps binary integrity (but with no guarantees that it is authentic), but your TLS client should do that anyway. I also don't want to give users a false hope that the SHA will protect them (since that's what they're used to, unfortunately).
